### ISSUE TITLE
Allow all ones Huffman codes

### ISSUE THREAD

**Issue by SimonSegerblomRex** [libjpeg-turbo/libjpeg-turbo#765]
**Clear and concise description of the issue:**
libjpeg-turbo refuses to decode images encoded with Huffman codes that are all ones. 

**Steps to reproduce (using *only* libjpeg-turbo):**
```sh
./djpeg -outfile out.pgm bogus.jpg
```

**Image needed in order to reproduce:**
![bogus](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/5e215061-73bf-4aa6-b279-c147c6a1c4cc)
(This is the first JPEG tile extracted from [this file](https://raw.pixls.us/getfile.php/4438/nice/Blackmagic%20-%20Blackmagic%20Pocket%20Cinema%20Camera%204k%20-%2012bit%20(16:9).dng).)

**Expected behavior:**
libjpeg-turbo should decode the image without complaining about the Huffman codes.

**Observed behavior:**
libjpeg-turbo refuses to decode the image with the message
```sh
Bogus Huffman table definition
```
due to [this check](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/jdhuff.c#L201) (since the last Huffman code is all ones).

**libjpeg-turbo release(s), commit(s), or branch(es) in which the bug was observed (always test the tip of the main branch or the latest [stable pre-release](https://libjpeg-turbo.org/DeveloperInfo/PreReleases) to verify that the bug hasn't already been fixed):**
0fc7313e545a3ff499c19ee6591bb87f0ad8b2a4

**Suggested fix:**
```diff
diff --git a/jdhuff.c b/jdhuff.c
index cd8c0847..ea2a0b0f 100644
--- a/jdhuff.c
+++ b/jdhuff.c
@@ -195,10 +195,7 @@ jpeg_make_d_derived_tbl(j_decompress_ptr cinfo, boolean isDC, int tblno,
       huffcode[p++] = code;
       code++;
     }
-    /* code is now 1 more than the last code used for codelength si; but
-     * it must still fit in si bits, since no code is allowed to be all ones.
-     */
-    if (((JLONG)code) >= (((JLONG)1) << si))
+    if (((JLONG)code - 1) >= (((JLONG)1) << si))
       ERREXIT(cinfo, JERR_BAD_HUFF_TABLE);
     code <<= 1;
     si++;
```
By changing this the attached image gets decoded without any problem.

**Reasoning:**
I'm not sure if this should be regarded a bug or not. It all depends on how you interpret the recommendations in [Annex K, section K.2](https://www.w3.org/Graphics/JPEG/itu-t81.pdf) regarding the generation of custom Huffman codes:

> Before starting the procedure, the values of FREQ are collected for V = 0 to 255 and the FREQ value for V = 256 is set to
1 to reserve one code point. [...] Reserving one code point guarantees that no code word can ever be all “1” bits.

at the same time it's stated that:

> This annex does **not** form an integral part of this Recommendation | International Standard

[libjpeg-turbo has decided to follow](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/jchuff.c#L963) the guideline in Annex K for the encoder which is probably a good decision, but the question is if it should really be enforced in the decoder. 

I'm not sure how widespread this issue is with JPEG encoders, but I'm afraid it might be common for Lossless JPEG encoded images (commonly encountered in DNG files). I just realized that I hadn't followed the guideline myself in my own simple python implementation of a Lossless JPEG encoder, and then the first image I downloaded from https://raw.pixls.us/#repo had the same issue.

(Not allowing the last Huffman code to be all ones in the encoder has a slight negative effect on the compression ratio, and I don't really understand why Annex K recommends against all ones codes. I haven't encountered any other Lossless JPEG decoders with any issues decoding these images.
**EDIT:** I guess the recommendation of not using an all ones Huffman code is to avoid the risk of unnecessary byte stuffing.)

**Comment by dcommander** [comment]
I would be a lot more receptive to enabling this just for lossless images, since we have encountered no examples of lossy JPEG images that require it.  However, I also need to understand how the image in question was generated.  Although it no longer causes an error when the aforementioned patch is applied, the image still doesn't appear to decode correctly.

I also need to understand whether this relates to #586.  It also appears that DCMTK did something similar, perhaps for the same reason: https://github.com/DCMTK/dcmtk/commit/760960589ce7d9bce913e825fa8ef05759ecdcd2.  I'll try to sort that out in the next few days.

**Comment by SimonSegerblomRex** [comment]
Thank you for the quick reply @dcommander! I agree that changing this requires some more research and there's no reason to rush a change.

> I would be a lot more receptive to enabling this just for lossless images, since we have encountered no examples of lossy JPEG images that require it.

Makes sense!

> However, I also need to understand how the image in question was generated.

I'm not sure, but I have a vague memory of Blackmagic using a version of the [lj92](https://github.com/ilia3101/MLV-App/blob/master/src/mlv/liblj92/lj92.c) Lossless JPEG encoder (only supporting predictor mode 6). I'll try to verify this.

> Although it no longer causes an error when the aforementioned patch is applied, the image still doesn't appear to decode correctly.

I think libjpeg-turbo decodes the image without issues after applying my patch. The reason it looks a bit strange is a common trick used when encoding bayer data with predictor mode 6: The JPEG image is encoded with double the width to align the color channels to get a good compression ratio. (Then the TIFF/DNG encoder just ignores the width and height information from the JPEG header.)

> I also need to understand whether this relates to #586.

I had a look at that issue, and it doesn't seem to be related.

> It also appears that DCMTK did something similar, perhaps for the same reason: [DCMTK/dcmtk@7609605](https://github.com/DCMTK/dcmtk/commit/760960589ce7d9bce913e825fa8ef05759ecdcd2). I'll try to sort that out in the next few days.

Interesting! Too bad the commit message isn't more informative.

I'll try to dig some more to see if I can find more affected DNG images and check if Adobe's encoder is affected.

**Comment by dcommander** [comment]
I had to do some reading and refresh my memory a bit.  You are correct that #586 is unrelated.  That issue involved some JPEG encoders that generate sub-optimal Huffman tables with unused out-of-range symbols.  libjpeg-turbo would never generate such images and rejects them, possibly because it cannot safely decode them without a buffer overrun.  (I haven't delved deeply into that, but suffice it to say that I don't think it's relevant to this issue, even though the error message was the same.)

Referring to https://github.com/libjpeg-turbo/libjpeg-turbo/issues/402#issuecomment-1353706456 and https://github.com/libjpeg-turbo/libjpeg-turbo/issues/402#issuecomment-1354352401, your proposed patch is essentially the same as the 2001-09-03 patch in DCMTK, the reason for which has been lost to history.  It is possible that the patch was necessitated by images generated with the Adobe DNG Converter, although [I could not personally reproduce that](https://github.com/libjpeg-turbo/libjpeg-turbo/issues/402#issuecomment-1349153637).  I intended to incorporate that patch for lossless mode if/when images were discovered that needed it.  Now that the images have been discovered, I just to need to know how they were generated so that I can document the scope of the issue.

**Comment by dcommander** [comment]
Wasn't the original image that failed created with Adobe's software?

**Comment by SimonSegerblomRex** [comment]
> Wasn't the original image that failed created with Adobe's software?

No, [the first example image ](https://raw.pixls.us/getfile.php/4438/nice/Blackmagic%20-%20Blackmagic%20Pocket%20Cinema%20Camera%204k%20-%2012bit%20(16:9).dng) seems to be have been created with some proprietary software as well. The EXIF data just says "6.1", I guess it's something from Blackmagic (but I don't know anything about this camera).

**Comment by SimonSegerblomRex** [comment]
I read the JPEG specification again and found this in Annex C:
> In addition, the codes shall be generated such that the all-1-bits code word of any length is reserved as a prefix for longer code words.

...so maybe these images actually shouldn't be treated as valid JPEG files.

**Comment by dcommander** [comment]
@SimonSegerblomRex I'm honestly not sure.  My read of Annex C, F.1.2.1.1, and F.1.2.2.1 is that a code consisting of all 1's is disallowed except as a "prefix for longer code words."  However, it also doesn't appear as if the code in question is actually used. Thus, this may actually be similar to #586 after all, except for the fact that #586 involves an unused code that is technically allowed in the spec, whereas this issue involves an unused code that seems to be disallowed.  The [other JPEG-1 reference implementation](https://github.com/thorfdbg/libjpeg) decodes the image in question without complaints.

I'm reasonably sure that a code of all 1's or a code > 16 would break our fast Huffman decoder.  However, only the slow Huffman decoder is used with lossless JPEG images.  It is possible that the slow Huffman decoder would catch such errant codes if they are actually used in the image (because of [this check](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/97772cba65aec15f8a7397f86131ffd63c1266d0/jdhuff.c#L465-L468)), but I'm not 100% sure of that.  Before I relax the bounds checks on Huffman table entries in lossless JPEG images, I would need to fully understand the ramifications to our specific implementation, including whether the out-of-range codes would be handled properly if they are actually used in the image.

@thorfdbg Would love to hear your take on this.

**Comment by thorfdbg** [comment]
> @thorfdbg <https://github.com/thorfdbg> Would love to hear your take on 
> this.
Annex K is informative and thus has nothing to say on normative 
requirements for Huffman codes, but Annex C is. The way how I read the 
paragraph 2 in this annex is that it disallows such codewords, and that 
it disallows also codewords longer than 16 bits.

Probably ISO would not let this text pass by today's standards (it 
formulates an encoder-side requirement and not a condition on a 
bitstream), but such formalities aside, I believe it seems quite explicit.

HTHH,

Thomas

**Comment by dcommander** [comment]
I thought so as well but wanted to make sure I wasn't misreading it.

My response to this will echo my response to #586:  The design of our Huffman decoder makes it impossible to know ahead of time whether a code in the Huffman table will actually be used, so `jpeg_make_d_derived_tbl()` must assume that all codes in the table will be used.  Thus, it must also reject codes that might cause the decoder to overrun/underrun one of its buffers/arrays or otherwise fail in a potentially exploitable manner.  In order to allow those codes, I would need to be 100% confident that the Huffman decoder will fail gracefully if the codes are actually encountered in the image.

Additionally, I think it would be a bad idea for an ISO/ITU-T reference implementation to allow a Huffman code that is clearly disallowed by the spec.  If the image is encoded incorrectly, then silently accepting it doesn't actually improve compatibility.  It just lowers the compatibility bar, allowing other encoders to get away with similar mistakes.  (That's why we have a spec and why the scope of it must not be allowed to creep.)  When libjpeg was created, the JPEG standard was brand new, and JFIF wasn't a formal standard at all.  Implementations of the brand new standard were all over the map, so one of Tom Lane's stated goals was to encourage industry convergence around a common JPEG file format by providing liberally-licensed free software that (strictly) implements the format.  It would be antithetical to that goal, not to mention the goals of the standards bodies, to allow non-conformant JPEG images to be decoded.

Downstream integrators, such as DCMTK, that are focused on compatibility with legacy lossless JPEG encoders (even incorrect ones) have shown willingness to apply downstream patches (e.g. [this one](https://github.com/DCMTK/dcmtk/commit/497043a84df12b85e736c1228f73e0f919926de0)) that relax the Huffman table checks.  The understanding is that such downstream software assumes the risks, including potential security risks, of relaxing those checks, because the truth is that I don't really know how our Huffman decoder will respond if the codes in question are actually used.  Without knowing that, I could not in good conscience allow such a patch upstream, even if it was disabled by default.

Honestly, I think the best solution would be a utility that optimizes the Huffman tables of lossless JPEG images that have unused codes.  You can achieve that currently using https://github.com/thorfdbg/libjpeg.git, but it requires a full decompress/recompress (not really an issue, since we're dealing with lossless JPEG images.)  Before anyone asks, jpegtran could not be repurposed for that, since it is built on an interface that deals only with DCT coefficients.

I am open to new arguments and data that might change my opinion of this, but for now, I am closing it as "won't implement".

**Comment by SimonSegerblomRex** [comment]
> I'm reasonably sure that a code of all 1's or a code > 16 would break our fast Huffman decoder. However, only the slow Huffman decoder is used with lossless JPEG images. It is possible that the slow Huffman decoder would catch such errant codes if they are actually used in the image (because of [this check](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/97772cba65aec15f8a7397f86131ffd63c1266d0/jdhuff.c#L465-L468)), but I'm not 100% sure of that.

The all ones Huffman code (4095, 12 bits) isn't used in the Blackmagic file, but the one in the Pentax file (1023, 10 bits) is actually used and libjpeg-trubo has no problem decoding the image (after bypassing the check for valid Huffman codes).

Without having looked into the details of the actual implementation in libjpeg-turbo I don't see any reason why the decoder should have any issue decoding all ones Huffman codes in the bit stream.

The only real problem is that the JPEG specification explicitly disallows all ones Huffman codes in the Huffman table definition due to the risk of FF bytes appearing there, but I see some problems with this:
1. The specification doesn't disallow having Huffman codes with all ones in the MSB, e.g. 0b1111111111111110 16 bits (0xFFFE), so the decoder needs to be careful with the risk of FF bytes (that shouldn't be treated as markers) appearing in the Huffman table definition anyway. (Or am I missing something here?)
2. What about all ones Huffman codes using less bits than 8? Then there's no risk of FF bytes in the table definition. What's the reason for not allowing them?
3. What about binary images? Not allowing the use of Huffman code 1 (1 bit) has a huge negative impact on the compression ratio.

This issue turned out to be a lot more complicated than I expected and I'm sorry for wasting so much of your time on it! I agree with @thorfdbg that the specification clearly disallows these Huffman codes, but the reasoning for it is vague. If it's really about the risk of FF codes in the table definition they should have come up with some other way around that.

Here's just one example of an image encoded using the Huffman code 127 (7 bits):
![huffman_code_127_with_7_bits](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/7ae8c735-fd72-4fc7-96e7-a2ca1e848cf6)
and the same image encoded using 255 (8 bits):
![huffman_code_254_with_8_bits](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/5f861828-456b-4297-b3eb-b8d941fc8081)
Using the all ones bits for the largest Huffman code for this image you...
* ...save one bit every time encoding SSSS=11 saving 1439 bits in total,
* ...**decrease** the byte stuffing from 3092 bytes to 3087 bytes. (Very small difference and probably varies from image to image, but still good to know that it at least doesn't increase the byte stuffing significantly (since I guess this must have been one of the original assumptions when writing the specification).

For a binary image you go from 0 bytes byte stuffing with valid Huffman codes (0b10 (2 bits) + 0b0 (1 bit)) to actually having to use byte stuffing with invalid Huffman codes (0b1 (1 bit) + 0b0 (1 bit)), but in my tests the penalty of being forced to use 2 bits instead of 1 is much worse than the increased byte stuffing.

(The images attached here were just created using a simple python script to demonstrate this difference.)

**Comment by dcommander** [comment]
Sorry, I guess I accidentally edited your comment above rather than replying to it.  Now GitHub won't let me delete my revision.  Ugh.

**Comment by SimonSegerblomRex** [comment]
> Update: I seem to have mostly restored it (not as easy as it should have been.)

Thanks, I just fixed the broken attachment links!

> FF bytes should always have a stuffed 0 byte after them to avoid such confusion.

Interesting! I always interpreted the spec in the way that byte stuffing only applies for the compressed image data (and not for the Huffman table definition).

**Comment by dcommander** [comment]
> Interesting! I always interpreted the spec in the way that byte stuffing only applies for the compressed image data (and not for the Huffman table definition).

It isn't an issue in the DHT segment (where the Huffman tables are defined), because the tables are of a specified length.  It's only an issue in the encoded data, and mainly because you can have restart markers interspersed.

**Comment by SimonSegerblomRex** [comment]
> > Interesting! I always interpreted the spec in the way that byte stuffing only applies for the compressed image data (and not for the Huffman table definition).
> 
> It isn't an issue in the DHT segment (where the Huffman tables are defined), because the tables are of a specified length. It's only an issue in the encoded data, and mainly because you can have restart markers interspersed.
> 
> Basically, I think you misinterpreted the spec. All 1's Huffman codes aren't disallowed because of potential confusion with markers. They are disallowed because all 1's is a special value that means something else.

Thanks! Yes, I think the spec is very confusing. 

In F.1.2.1.1 we have
> The Huffman codes for the difference categories are generated in such a way that no code consists entirely of 1-bits
(X’FF’ prefix marker code avoided).

but there's no reason given for why the FF prefix marker code should be avoided (and this is not even true for code words using less than 8 bits).

In Annex C there is
> In addition, the codes shall be generated such that the all-1-bits code word of any length is reserved as a prefix for longer code words.

but again this doesn't make any sense for a custom Huffman table tied to a specific image. There's no reason to reserve code words that are not used. (Why would you for instance reserve the code word 0b11 (2 bits) when encoding a binary image?)

> My biggest concern would be with a file that uses an all 1's Huffman code with full sample precision, e.g. a 12-bit file that uses a code of 4095.

For Lossless JPEG encoded bit streams the sample precision doesn't really affect the decoding (or encoding) process, so I don't think there's anything special with the code word 4095 (12 bits) (but I realize that there's more work required to establish this).

I'll see if I get back to this issue some day. It would be nice to be able to use libjpeg-turbo as the decoder for Lossless JPEG encoded DNG files in open-source RAW editors like Rawtherapee and Darktable. A small patch like this would be easy enough to maintain in a fork, so I don't think that this is a showstopper. Thank you for all your time spent discussing this!

**Comment by thorfdbg** [comment]
For the records, I added now a check in the ISO reference code, and it now generates a warning if it finds such a codeword. It proceeds to decode, though.

**Comment by dcommander** [comment]
@thorfdbg Can you comment as to why the all 1's Huffman code word is reserved?

**Comment by dcommander** [comment]
> For Lossless JPEG encoded bit streams the sample precision doesn't really affect the decoding (or encoding) process, so I don't think there's anything special with the code word 4095 (12 bits) (but I realize that there's more work required to establish this).

I'm just speaking from my experience with the specific Huffman decoder implementation in libjpeg[-turbo]. I don't know for a fact that it would choke on Huffman codes with all 1's under certain circumstances, nor do I know for a fact that it would choke on Huffman codes > 16, but I could envision either scenario causing a table overrun-- especially with 16-bit precision, in which case the sample data type is only barely big enough to hold the maximum code. These are gut feelings based on my experience. Again, I would need to be convinced that such is mathematically impossible with our specific implementation.

As to your other questions, I have no idea.

**Comment by thorfdbg** [comment]
I would love to, but I do not know either where that restriction is coming from. I'm second generation JPEG, first generation is retired by now. Anyhow, a standard does not need to explain why...

I can only guess on the reasons... With such a Huffman code present, very long runs of FFs are possible, and thus the bitstream can expand by a factor of two due to byte stuffing. Probably people wanted to limit the maximal size a JPEG codestream can reach, and thus the maximal bandwidth required for a receiver.

**Comment by thorfdbg** [comment]
The maximum bitsize of 16 for Huffman codes is, however, easily explained. A typical hardware implementation of a Huffman decoder is just a shift register and a comparator, and the reconstructed symbol is found by comparing the contents of the shift register with a lookup table that contains the symbol and the bitsize. The bitsize then defines how many new bits are shifted from the codestream into the register. By limiting the bitsize of the code, the size of the shift register remains limited, and thus the hardware remains simpler.

**Comment by dcommander** [comment]
That makes total sense and also answers @SimonSegerblomRex's question of why all-1's codes less than 8 bits aren't allowed. (Because the codes are packed together across byte boundaries, a long run of all 1's codes would produce a lot of sequential 0xFF bytes, regardless of the number of bits in the codes.)

So basically, the conclusion is: yes, all 1's codes are disallowed. I would be willing to relax the restrictions on table entries via a compile-time option (thus allowing downstream software to easily work around this and #586) but only if I can develop confidence that doing so could never cause a buffer overrun in our implementation.

**Comment by Melirius** [comment]
Effectively all-1 codes (besides the longest one, 16-bit in JPEG case) are forbidden by Huffman encoding scheme. If there will be, for example, two all-1 codes of 7 and 8 bits, they are indistinguishable by decoder: is this sequence of 8 1-bits a 7-bit code plus the first 1-bit of value, or just 8-bit code?

**Comment by SimonSegerblomRex** [comment]
>  If there will be, for example, two all-1 codes of 7 and 8 bits, they are indistinguishable by decoder: is this sequence of 8 1-bits a 7-bit code plus the first 1-bit of value, or just 8-bit code?

Yes, that's true regarding Huffman codes, but that's not what's been discussed here. There can obviously only be one Huffman code with all ones (the largest one) per Huffman table, but the JPEG specification prohibits the use of these codes (including the 16 bit one). 

Pentax is one example of a vendor that has used a JPEG encoder not following the specification in this regard. The reason for that might be to get slightly better compression ratios, or they simply missed this quirk in the specification. Here's a comment in [libjpeg](https://github.com/thorfdbg/libjpeg/blob/e5681eeeae41ea773b86610f2a42a82a83a8dbd0/coding/huffmancoder.cpp#L73):
```c++
        // Note: This test also checks that there is no code with all 1 bits set
        // This is what Annex K.2 of Recommendation ITU-T T.81 (1992) | ISO/IEC 10918-1:1994
        // says, and the introduction of Annex C enforces. There is, from a pure coding
        // perspective, no need to enforce this rule, i.e. a all-1 codeword would still
        // create a valid prefix-free code.
```

A bug free JPEG decoder has no problem decoding these files and libjpeg-turbo is probably the first Lossless JPEG decoder refusing to decode them. If we can rule out that there are no bugs in libjpeg-turbo it should be fine to enable this feature (at least through a compiler flag and with a warning), but I'm not sure if it's worth the effort only to support decoding Pentax files. I'll update this issue if I find more example images.

**Comment by sigurdle** [comment]
Following up on this (closed) issue — I'd like to ask you to reconsider, but with a **narrower, decoder-only proposal** than the original, plus a concrete real-world file and a precedent that's already in the code.

### The problem, concretely

libjpeg-turbo's **lossless** decoder aborts with `JERR_BAD_HUFF_TABLE` ("Bogus Huffman table definition") on lossless JPEGs whose Huffman table uses the all-ones codeword — which real cameras emit. Example: a Canon EOS 5D Mark III DNG (lossless, SOF3, 14-bit, predictor 6) from the CC0 sample set at [raw.pixls.us](https://raw.pixls.us/) (getfile id 2204). Its DC/lossless Huffman table:

```
BITS    = 1 0 2 2 2 3 1 1 1 2 0 0 0 0 0 0
HUFFVAL = 3 4 2 5 1 6 0 7 8 9 10 11 12 13 0
```

Canonical code generation assigns, at length 10, the codeword `0b1111111111` (all ones), so `jpeg_make_d_derived_tbl()` computes `code = 1024 >= (1 << 10)` and calls `ERREXIT(..., JERR_BAD_HUFF_TABLE)`. dcraw, LibRaw and other lossless decoders read this file without issue.

### Minimal reproducer (~50 bytes)

SOF3, 8-bit, 4×1, one component; a 2-symbol length-1 table whose canonical codes are `0` and `1` (the `1` is the all-ones codeword). It decodes to samples `128,129,128,128`:

```
FF D8
FF C4 00 15 00  02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  00 01
FF C3 00 0B 08 00 01 00 04 01 01 11 00
FF DA 00 08 01 01 00 01 00 00
73
FF D9
```

### Why this is worth a second look — and different from before

1. **Decoder-only, lossless-scoped.** The earlier patch relaxed the check for *all* modes. This gates on `cinfo->master->lossless`, so baseline and progressive decoding are byte-for-byte unchanged.
2. **Precedent already in this function.** `jpeg_make_d_derived_tbl()` *already* special-cases lossless a few lines below — the DC symbol-range check accepts `0..16` "in lossless mode" vs `0..15` otherwise (`cinfo->master->lossless ? 16 : 15`). Lossless already carries its own table-validation rules here.
3. **It's what lossless support is for.** Real-world lossless JPEG is overwhelmingly camera raw (DNG/CR2) and DICOM, and those encoders routinely produce complete Huffman codes.
4. **These are valid prefix codes, not corruption.** A complete code that uses the all-ones codeword is unambiguously decodable; the Annex K "reserve all-ones" rule is *encoder* guidance. Byte-stuffing (`FF`→`FF 00`) is handled at the byte layer, independent of code values.

### Proposed change (`jdhuff.c`, `jpeg_make_d_derived_tbl`)

```diff
-    if (((JLONG)code) >= (((JLONG)1) << si))
+    if (((JLONG)code) > (((JLONG)1) << si) ||
+        (((JLONG)code) == (((JLONG)1) << si) && !cinfo->master->lossless))
       ERREXIT(cinfo, JERR_BAD_HUFF_TABLE);
```

**This is already being patched downstream.** [Tom94's libjpeg-turbo fork](https://github.com/Tom94/libjpeg-turbo) carries commits titled *"fix(lossless): accept all-ones huffman"* ("Needed for some DNGs", referencing this issue), and [SimonSegerblomRex/imagecodecs](https://github.com/SimonSegerblomRex/imagecodecs) has parallel lossless-DNG work — so people are independently forking to get this. Those use the **global** form (`code - 1 >= 1 << si`); the version above is the narrower **lossless-only** variant, so baseline and progressive are left untouched.

With this applied: the Canon DNG decodes, its output is bit-identical to an independent from-scratch lossless decoder, and the baseline/progressive/CMYK regression suite is unchanged. Happy to open a PR with a test if you're open to it. Thanks for considering.

---
*Disclosure: this investigation and patch were prepared with the help of an AI coding assistant (Claude). Everything above was verified on a real patched libjpeg-turbo 3.1.1 build — the Canon DNG fails on stock with `JERR_BAD_HUFF_TABLE` and decodes with the patch (output bit-identical to an independent lossless decoder), and the baseline/progressive/CMYK regression suite is unchanged.*

**Comment by dcommander** [comment]
@sigurdle This code base was created by human beings and is maintained by a single underpaid human being.  I am extremely unreceptive to AI-assisted arguments, especially when they ignore the dozens of human comments above that lay out the case against including the proposed enhancement.  Per above, there are valid concerns about whether opening up the Huffman decoder in the proposed manner would create a potential for overrunning the DHT via a specially crafted malformed lossless JPEG image.  Convince me-- using your brain, not AI-- that that can't happen mathematically.  Also post a link to the actual DNG image you are referring to.  I don't know what "getfile id 2204" means.

**Comment by sigurdle** [comment]
Here's the link to a dng file that your reader fails on
https://raw.pixls.us/getfile.php/2204/nice/Canon-EOS-5D-Mark-III.dng
I tried to attach, but it wouldnt allow dng files

As for using my brain to prove that the change is security safe, no i'm not smart enough unfortunately, and also not willing to spend the time digging into someone elses source code. I did have the AI check if security was compromised. Then it said:
"Why it can't overrun." and laid out the reasons, but im not going to paste that since you dont like AI proof, I can provide it if you want, but its not things I myself understand, since i'm not familiar with your code.

I have the change in my own code, so I can load dng files. On a personal level it's not important to me if you put change in your code or not, I thought I was doing you a favor by submitting this bug report. But of course I understand non-receptiveness to AI.. Not sure what to do about that. I myself can't convince you with my brain.

Best regards,
Sigurd Lerstad

**Comment by dcommander** [comment]
@sigurdle The problem is that these AIs don't really understand anything.  They're just stringing words together based on massive probabilistic language models.  In this case, the models they're using are trained on the entire Internet, so they likely include some of my code, documentation, commit logs, comments, forum messages, etc.  There is no way to know how the AI arrived at its answer, and it is well documented that AIs are biased toward yes answers (i.e. telling you what you want to hear.)  A representative for the relevant standards body participated in the discussion above, and I believe we definitively concluded that all-ones Huffman codes are disallowed by the JPEG-1 spec, although the reasons why are lost to history.  Your comment did not address or even account for that previous discussion.

Note that one of the downstream codecs you cite is maintained by the original poster of this issue, who re-encoded his images to conform to the spec.  That's the real issue with this: the fact that we want to discourage propagation of non-conformant JPEG images.  That was, in fact, one of the reasons why Tom Lane created libjpeg: to steer the community in the direction of a common conformant JPEG variant by providing a conformant, ubiquitous codec that is license-compatible with all OSS and proprietary applications.  However, Tom stopped working on libjpeg before lossless support could be included, so there was no ubiquitous codec to steer the community in the direction of a common conformant lossless JPEG variant until libjpeg-turbo 3.0 was released 3 years ago.  Thus, there are still some non-conformant implementations out there.

We have established that the Huffman codes in question are non-conformant, so the question becomes whether to let that slide, which would encourage the propagation of such non-conformant images, or to continue to draw a line in the sand per Tom Lane's original vision.  I chose the latter.  I considered making this a compile-time option that is disabled by default, so downstream codecs that are willing to accept the risk could allow such non-conformant images, but that in and of itself would still be an implicit endorsement of those images (albeit a softer endorsement than supporting them by default.)  I fully understand that makers of consumer software might need to be slightly more liberal and allow slightly non-conformant images, but as a reference implementation of JPEG-1, I believe strongly that libjpeg-turbo should draw a line in the sand and do nothing to encourage non-conformance.
